mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/impl/QueueOfferState.scala

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.mqtt.streaming.impl
+
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.Behaviors
+import akka.stream.QueueOfferResult
+
+private[mqtt] object QueueOfferState {
+
+  /**
+   * A marker trait that holds a result for SourceQueue#offer
+   */
+  trait QueueOfferCompleted {
+    def result: Either[Throwable, QueueOfferResult]
+  }
+
+  /**
+   * A behavior that stashes messages until a response to the SourceQueue#offer
+   * method is received.
+   *
+   * This is to be used only with SourceQueues that use backpressure.
+   */
+  def waitForQueueOfferCompleted[T](behavior: Behavior[T], stash: Seq[T]): Behavior[T] =
+    Behaviors
+      .receive[T] {
+        case (context, completed: QueueOfferCompleted) =>
+          completed.result match {
+            case Right(QueueOfferResult.Enqueued) =>
+              stash.foreach(context.self.tell)
+
+              behavior
+
+            case Right(other) =>
+              throw new IllegalStateException(s"Failed to offer to queue: $other")
+
+            case Left(failure) =>
+              throw failure
+          }
+
+        case (_, other) =>
+          waitForQueueOfferCompleted(behavior, stash :+ other)
+      }
+      .orElse(behavior) // handle signals immediately
```

Comment:
One thing I'm not clear on are the semantics here. Note that on L43, we catch all events that flow. Thus, the intention is that all signals go to `behavior` but nothing else (until we hit L34 in the future)

If `behavior`'s signal handler (i.e. the behavior being run via `orElse`) returns `Behavior.same`, I'd like to clarify that it doesn't change the behavior of the actor to itself, i.e. `waitForQueueOfferCompleted`  (L26-L46) remains the behavior. Hopefully my question makes sense..

Reply:
I *think* the behaviour would indeed remain the same i.e. stay as per the behaviour of `waitForQueueOfferCompleted`. But I've not yet proved this.

However, given L46 with the `orElse`, I'm not seeing the signal handler of `behavior` being applied. When I run the tests I see a `DeathPactException` which means that there is no signal handler for `Terminated`, but there is such a signal handler for `behavior`...